README.md
# Artstor Viewer, Angular Component
Media viewer for content within Artstor Image Workspace

## Ithaka Open Source Initiative
This repository is open sourced by Ithaka as part of our initiative to increase transparency and contribute to the community. This particular repository is a working repository not intended as software available for consumption.

[Learn more](http://artstor.org/open-source) about our open source initiative

**Copyright 2018 Ithaka Harbors, Inc.**

## Building the Project

_Commands use `yarn`, which we recommend, but can be replaced with `npm`_

Using this repo:
- `yarn start` to run a live-reload server with the demo app
- `yarn run test` to test in watch mode, or `yarn run test:once` to only run once
- `yarn run build` to build the library
- `yarn run lint` to lint 
- `yarn run clean` to clean
- `yarn run integration` to run the integration e2e tests
- `yarn install ./relative/path/to/lib` after `yarn run build` to test locally in another app

To publish a new version:
- `yarn release` to increment version
- `git commit` to commit the version
- `yarn build` to build with new version
- `npm publish dist` to publish to yarn/npm
- `git push` to push version to master

## The Demo App
Used to test the viewer, and provides multiple asset ids to test different use cases.

1. [Clone](#clone "Clone it from github") or [download](#download "download it from github") the **artstor-viewer**.
1. Install npm packages with `yarn install`.
1. Run `npm start` to launch the sample application.

## Directory
```
src/
├── demo/
|  └── app/
|     ├── app.component.ts
|     └── app.module.ts
└── lib/
   ├── index.ts
   └── src/
      ├── component/
      |  └── artstor-viewer.component.ts
      ├── service/
      |  └── lib.service.ts
      └── module.ts

```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.3.7"></a>
## [1.3.7](https://github.com/ithaka/ng-artstor-viewer/compare/v1.3.6...v1.3.7) (2019-01-25)



<a name="1.3.6"></a>
## [1.3.6](https://github.com/ithaka/ng-artstor-viewer/compare/v1.3.5...v1.3.6) (2019-01-08)



<a name="1.3.5"></a>
## [1.3.5](https://github.com/ithaka/ng-artstor-viewer/compare/v1.3.4...v1.3.5) (2018-12-04)



<a name="1.3.4"></a>
## [1.3.4](https://github.com/ithaka/ng-artstor-viewer/compare/v1.3.3...v1.3.4) (2018-11-28)



<a name="1.3.3"></a>
## [1.3.3](https://github.com/ithaka/ng-artstor-viewer/compare/v1.3.1...v1.3.3) (2018-11-09)



<a name="1.3.2"></a>
## [1.3.2](https://github.com/ithaka/ng-artstor-viewer/compare/v1.2.6...v1.3.2) (2018-11-07)



<a name="1.3.1"></a>
## [1.3.1](https://github.com/ithaka/ng-artstor-viewer/compare/v1.2.5...v1.3.1) (2018-11-06)



<a name="1.2.6"></a>
## [1.2.6](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.1...v1.2.6) (2018-11-06)



<a name="1.2.5"></a>
## [1.2.5](https://github.com/ithaka/ng-artstor-viewer/compare/v1.2.4...v1.2.5) (2018-10-18)



<a name="1.2.4"></a>
## [1.2.4](https://github.com/ithaka/ng-artstor-viewer/compare/v1.2.3...v1.2.4) (2018-10-16)



<a name="1.2.3"></a>
## [1.2.3](https://github.com/ithaka/ng-artstor-viewer/compare/v1.2.2...v1.2.3) (2018-10-16)



<a name="1.2.2"></a>
## [1.2.2](https://github.com/ithaka/ng-artstor-viewer/compare/v1.2.1...v1.2.2) (2018-10-11)



<a name="1.2.1"></a>
## [1.2.1](https://github.com/ithaka/ng-artstor-viewer/compare/v1.2.0...v1.2.1) (2018-10-09)



<a name="1.2.0"></a>
# [1.2.0](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.12...v1.2.0) (2018-10-09)



<a name="1.1.13"></a>
## [1.1.13](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.12...v1.1.13) (2018-09-27)



<a name="1.1.12"></a>
## [1.1.12](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.11...v1.1.12) (2018-09-19)



<a name="1.1.11"></a>
## [1.1.11](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.10...v1.1.11) (2018-09-13)



<a name="1.1.10"></a>
## [1.1.10](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.9...v1.1.10) (2018-09-07)



<a name="1.1.9"></a>
## [1.1.9](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.8...v1.1.9) (2018-08-31)



<a name="1.1.8"></a>
## [1.1.8](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.6...v1.1.8) (2018-08-15)



<a name="1.1.7"></a>
## [1.1.7](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.4...v1.1.7) (2018-08-06)



<a name="1.1.6"></a>
## [1.1.6](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.5...v1.1.6) (2018-08-03)



<a name="1.1.5"></a>
## [1.1.5](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.4...v1.1.5) (2018-08-03)



<a name="1.1.4"></a>
## [1.1.4](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.37...v1.1.4) (2018-07-24)



<a name="1.1.3"></a>
## [1.1.3](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.2...v1.1.3) (2018-05-14)



<a name="1.1.2"></a>
## [1.1.2](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.1...v1.1.2) (2018-05-09)



<a name="1.1.1"></a>
## [1.1.1](https://github.com/ithaka/ng-artstor-viewer/compare/v1.1.0...v1.1.1) (2018-05-09)



<a name="1.0.49"></a>
## [1.0.49](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.48...v1.0.49) (2018-05-03)



<a name="1.0.48"></a>
## [1.0.48](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.47...v1.0.48) (2018-05-03)



<a name="1.0.47"></a>
## [1.0.47](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.46...v1.0.47) (2018-04-20)



<a name="1.0.46"></a>
## [1.0.46](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.45...v1.0.46) (2018-04-16)



<a name="1.0.45"></a>
## [1.0.45](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.44...v1.0.45) (2018-04-04)



<a name="1.0.44"></a>
## [1.0.44](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.43...v1.0.44) (2018-04-04)



<a name="1.0.43"></a>
## [1.0.43](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.42...v1.0.43) (2018-03-19)



<a name="1.0.42"></a>
## [1.0.42](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.41...v1.0.42) (2018-03-13)



<a name="1.0.41"></a>
## [1.0.41](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.40...v1.0.41) (2018-03-09)



<a name="1.0.40"></a>
## [1.0.40](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.39...v1.0.40) (2018-03-09)



<a name="1.0.39"></a>
## [1.0.39](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.38...v1.0.39) (2018-03-07)



<a name="1.0.38"></a>
## [1.0.38](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.1...v1.0.38) (2018-03-07)



<a name="1.0.37"></a>
## [1.0.37](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.36...v1.0.37) (2018-03-01)



<a name="1.0.36"></a>
## [1.0.36](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.35...v1.0.36) (2018-03-01)



<a name="1.0.35"></a>
## [1.0.35](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.1...v1.0.35) (2018-03-01)



<a name="1.0.34"></a>
## [1.0.34](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.33...v1.0.34) (2018-02-20)



<a name="1.0.33"></a>
## [1.0.33](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.32...v1.0.33) (2018-02-20)



<a name="1.0.32"></a>
## [1.0.32](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.31...v1.0.32) (2018-02-20)



<a name="1.0.31"></a>
## [1.0.31](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.30...v1.0.31) (2018-02-16)



<a name="1.0.30"></a>
## [1.0.30](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.29...v1.0.30) (2018-02-16)



<a name="1.0.29"></a>
## [1.0.29](https://github.com/ithaka/ng-artstor-viewer/compare/v1.0.28...v1.0.29) (2018-02-15)



<a name="1.0.28"></a>

src/lib/src/asset.interface.ts
// Project Dependencies
import { AssetData, MetadataField, FileProperty, CollectionValue } from './component/artstor-viewer.component'

export class Asset {
    id: string
    groupId: string
    typeId: number // determines what media type the asset is
    typeName: string // the name correlating to the typeId
    title: string
    thumbnail_url: string
    thumbnail_size: number = 3
    kalturaUrl: string
    downloadLink: string
    downloadName: string
    tileSource: string | string[]
    collectionType: number
    contributinginstitutionid: number
    personalCollectionOwner: number
    // Not reliably available
    categoryId: string
    categoryName: string
    collections: CollectionValue[]
    collectionId: string
    collectionName: string
    SSID: string
    fileName: string
    updated_on: string

    viewportDimensions: {
        contentSize?: any,
        zoom?: number,
        containerSize?: any,
        center?: any
    } = {}

//   private dataLoadedSource = new BehaviorSubject<boolean>(false)
//   public isDataLoaded = this.dataLoadedSource.asObservable()

    public disableDownload: boolean = false

    /** Used for holding asset file properties array from the service response */
    filePropertiesArray: FileProperty[] = []
    /** Used for holding formatted asset metadata from the service response */
    formattedMetadata: FormattedMetadata = {}

    /** Used for holding media resolver info from the service response */
    viewerData?: {
        base_asset_url?: string,
        panorama_xml?: string
    }

    constructor(assetData: AssetData, testEnv ?: boolean) {
        if (!assetData) {
            throw new Error('No data passed to construct asset')
        }
        this.initAssetProperties(assetData, testEnv)
    }

    private formatMetadata(metadata: MetadataField[]): FormattedMetadata {
        let formattedData: FormattedMetadata = {}
        for (let data of metadata) {
            // this is stupid, but if there's a link then it needs to be assigned to the fieldValue
            if (data.link) {
                data.fieldValue = data.link
            }

            // if the field exists, add to it (make sure the field vaue exists)
            if (formattedData[data.fieldName] && data.fieldValue) {
                formattedData[data.fieldName].push(data.fieldValue)
            } else if(data.fieldValue) { // otherwise make a new field (make sure the field vaue exists)
                formattedData[data.fieldName] = [data.fieldValue]
            }
        }
        return formattedData
    }

    private buildDownloadLink(data: AssetData): string {
        let downloadLink: string
        switch (data.object_type_id) {
            case 20:
            case 21:
            case 22:
            case 23:
                // Non-image Download Link format: /media/ARTSTORID/TYPEID
                downloadLink = [data.baseUrl + 'media', this.id, data.object_type_id].join("/")
                break
            default:
                if (data.image_url) { //this is a general fallback, but should work specifically for images and video thumbnails
                    let imageServer = 'http://imgserver.artstor.net/' // TODO: check if this should be different for test
                    let url = imageServer + data.image_url + "?cell=" + data.download_size + "&rgnn=0,0,1,1&cvt=JPEG"
                    downloadLink = data.baseUrl + "api/download?imgid=" + this.id + "&url=" + encodeURIComponent(url)
                } else {
                    // nothing happens here because some assets are not allowed to be downloaded
                }
        }
        return downloadLink
    }

    /**
     * Sets the correct typeName based on a map
     * @param typeId the asset's object_type_id
     */
    private initTypeName(typeId: number): string {
        let objectTypeNames: { [key: number]: string } = {
            1: 'specimen',
            2: 'visual',
            3: 'use',
            6: 'publication',
            7: 'synonyms',
            8: 'people',
            9: 'repository',
            10: 'image',
            11: 'panorama',
            12: 'audio',
            13: '3d',
            21: 'powerpoint',
            22: 'document',
            23: 'excel',
            24: 'kaltura'
        }
        return objectTypeNames[typeId]
    }

    get creator(): string {
        return (this.formattedMetadata.Creator && this.formattedMetadata.Creator[0]) || ''
    }
    get date(): string {
        return (this.formattedMetadata.Date && this.formattedMetadata.Date[0]) || ''
    }
    get description(): string {
        return (this.formattedMetadata.Description && this.formattedMetadata.Description[0]) || ''
    }

    /**
     * Sets up the Asset object with needed properties
     * - Behaves like a delayed constructor
     * - Reports status via 'this.dataLoadedSource' observable
     */
    private initAssetProperties(data: AssetData, testEnv?: boolean): void {
        let storUrl: string = testEnv ? '//stor.stage.artstor.org' : '//stor.artstor.org' 
        // Set array of asset metadata fields to Asset, and format
        if (data.metadata_json) {
            this.formattedMetadata = this.formatMetadata(data.metadata_json)
        }
        this.id = data.object_id
        this.typeId = data.object_type_id
        this.title = data.title
        this.categoryId = data.category_id
        this.categoryName = data.category_name
        this.collections = data.collections
        this.collectionId = data.collection_id
        this.collectionName = data.collection_name
        this.filePropertiesArray = data.fileProperties
        this.collectionType = data.collection_type
        this.contributinginstitutionid = data.contributinginstitutionid
        this.personalCollectionOwner = data.personalCollectionOwner
        // we control the default size of the thumbnail url
        this.thumbnail_url = this.replaceThumbnailSize(data.thumbnail_url, this.thumbnail_size)
        this.typeId = data.object_type_id
        this.typeName = this.initTypeName(data.object_type_id)
        this.disableDownload =  data.download_size === '0,0'
        this.SSID = data.SSID
        // Set filename
        if (this.filePropertiesArray && this.filePropertiesArray[0]){
            let fileProperty = this.filePropertiesArray.find((obj) => {
                return !!obj && !!obj.fileName
            })
            this.fileName = fileProperty ? fileProperty.fileName : 'file'
        } else {
            this.fileName = 'file'
        }
        this.updated_on = data.updated_on
        // Set Download information
        let fileExt = this.fileName.substr(this.fileName.lastIndexOf('.'), this.fileName.length - 1)
        this.downloadName = this.title.replace(/\./g,'-') + '.' + fileExt
        this.downloadLink = this.buildDownloadLink(data)
        data.viewer_data && (this.viewerData = data.viewer_data)

        // Save the Tile Source for IIIF
        //  sometimes it doesn't come back with .fpx, so we need to add it
        let imgPath
        if (data.image_url.lastIndexOf('.fpx') > -1) {
            imgPath = '/' + data.image_url.substring(0, data.image_url.lastIndexOf('.fpx') + 4)
        } else {
            imgPath = '/' + data.image_url
        }
        if (data.image_compound_urls && data.image_compound_urls[0]) {
            for (let i = 0; i < data.image_compound_urls.length; i++) {
                let path = data.image_compound_urls[i]
                // path = path.replace('/info.json','')
                // data.image_compound_urls[i] = '//tsstage.artstor.org/rosa-iiif-endpoint-1.0-SNAPSHOT/fpx' + encodeURIComponent(path) + '/info.json'
                data.image_compound_urls[i] = storUrl + '/fcgi-bin/iipsrv.fcgi?IIIF=' + path
            }
            this.tileSource = data.image_compound_urls
        } else {
            this.tileSource = data.tileSourceHostname + '/rosa-iiif-endpoint-1.0-SNAPSHOT/fpx' + encodeURIComponent(imgPath) + '/info.json'
        }

        // set up kaltura info if it exists
        if (data.fpxInfo) {
            this.kalturaUrl = data.fpxInfo.imageUrl.replace('http://', '//')
        }
    }

    /**
     * Sometimes the thumbnail url will not exist and we need to fall back to a different thumbnail size
     */
    public fallbackThumbnailUrl(): void {
        if (this.thumbnail_size > 0) {
            this.thumbnail_size --
            this.thumbnail_url = this.replaceThumbnailSize(this.thumbnail_url, this.thumbnail_size)
        }
    }

    private replaceThumbnailSize(url: string, size: number): string {
        return url.replace(/(size)[0-9]/g, 'size' + size)
    }
}

export interface FormattedMetadata {
    [fieldName: string]: string[]
}